newlib/libc/include/assert.h

```diff
@@ -43,8 +43,8 @@ extern "C" {
 
 #undef assert
 
-#ifdef NDEBUG           /* required by ANSI standard */
-# define assert(__e) ((void)0)
+#ifdef NDEBUG
+# define assert(__e) ((__e) ? (void)0 : __unreachable())
```

Comment:
I do not think this is right. In release builds (when NDEBUG is defined), the assert code should be removed, so the original code is correct.

Also, why did you remove comment "required by ANSI standard". Is the comment wrong?

Reply:
Well, this slightly diverges from the ANSI standard now.
But I can see how this might have some undesirable consequences, maybe a separate `assume()` is a better solution here.

Reply:
I can see benefits to leaving the code visible to the compiler so that it gets validated. However, the C standard is prescriptive about the assert macro:

> the assert macro is defined simply as
> ```
> #define assert(ignore) ((void)0)
> ```

I don't think we can vary from that and still remain standards compliant.